File: python/496. Next Greater Element I.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        st = []
        n1, n2 = len(nums1), len(nums2)

        nxt = [-1 for _ in range(n2)]
        for i, num in enumerate(nums2):
            while st and nums2[st[-1]] < num:
                nxt[st.pop()] = num
            st.append(i)
        
        res = []
        for i, num in enumerate(nums1):
            j = nums2.index(num)
            res.append(nxt[j])
        return res
        

File: go/src/leetcode/496. Next Greater Element I/496. Next Greater Element I.go
/**
@author ZhengHao Lou
Date    2021/10/26
*/
package main

import (
	"fmt"
)

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	n2 := len(nums2)
	t := make([]int, n2)
	for i := range nums2 {
		t[i] = -1
	}

	st := []int{}
	for i, num := range nums2 {
		for len(st) > 0 && nums2[st[len(st) - 1]] < num {
			k := st[len(st) - 1]
			st = st[:len(st) - 1]
			t[k] = num
		}
		st = append(st, i)
	}
	fmt.Println(t)

	mapper := make(map[int]int)
	for i, num := range nums2 {
		mapper[num] = i
	}
	next := []int{}
	for _, num := range nums1 {
		i := mapper[num]
		next = append(next, t[i])
	}
	return next
}

func main() {
	fmt.Println(nextGreaterElement([]int{4,1,2}, []int{1,3,4,2}))
	fmt.Println(nextGreaterElement([]int{2,4}, []int{1,2,3,4}))
}
